xcvm/cosmwasm/contracts/interpreter/src/contract.rs

```diff
@@ -0,0 +1,288 @@
+	let registry_address = deps.api.addr_validate(&msg.registry_address)?;
+	let config = Config { registry_address };
+	CONFIG.save(deps.storage, &config)?;
+
+	Ok(Response::new().set_data(to_binary(&(msg.network_id.0, msg.user_id))?))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+	deps: DepsMut,
+	env: Env,
+	info: MessageInfo,
+	msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+	match msg {
+		ExecuteMsg::Execute { program } => interpret_program(deps, env, info, program),
+	}
+}
+
+pub fn interpret_program(
+	mut deps: DepsMut,
+	_env: Env,
+	_info: MessageInfo,
+	program: XCVMProgram,
+) -> Result<Response, ContractError> {
+	let mut response = Response::new();
+
+	for instruction in program.instructions {
+		response = match instruction {
+			Instruction::Call { encoded } => interpret_call(encoded, response),
+			Instruction::Spawn { network, salt, assets, program } =>
+				interpret_spawn(network, salt, assets, program, response),
+			Instruction::Transfer { to, assets } =>
+				interpret_transfer(&mut deps, to, assets, response),
+		}?;
+	}
+	Ok(response)
+}
+
+pub fn interpret_call(encoded: Vec<u8>, response: Response) -> Result<Response, ContractError> {
+	let cosmos_msg: CosmosMsg =
+		serde_json_wasm::from_slice(&encoded).map_err(|_| ContractError::InvalidCallPayload)?;
+
+	Ok(response.add_message(cosmos_msg))
+}
+
+pub fn interpret_spawn(
+	network: NetworkId,
+	salt: Vec<u8>,
+	assets: Funds,
+	program: XCVMProgram,
+	response: Response,
+) -> Result<Response, ContractError> {
+	#[derive(Serialize)]
+	struct SpawnEvent {
+		network: NetworkId,
+		salt: Vec<u8>,
+		assets: Funds,
+		program: XCVMProgram,
+	}
+
+	let data = SpawnEvent { network, salt, assets, program };
+
+	Ok(response.add_attribute(
+		"spawn",
+		serde_json_wasm::to_string(&data).map_err(|_| ContractError::DataSerializationError)?,
+	))
+}
+
+pub fn interpret_transfer(
+	deps: &mut DepsMut,
+	to: String,
+	assets: Funds,
+	mut response: Response,
+) -> Result<Response, ContractError> {
+	let config = CONFIG.load(deps.storage)?;
+	let registry_addr = config.registry_address.into_string();
+
+	for (asset_id, amount) in assets.0 {
+		let query_msg = AssetRegistryQueryMsg::GetAssetContract(asset_id.into());
+
+		let cw20_address: GetAssetContractResponse = deps.querier.query(
+			&WasmQuery::Smart { contract_addr: registry_addr.clone(), msg: to_binary(&query_msg)? }
+				.into(),
+		)?;
+		let contract = Cw20Contract(cw20_address.addr.clone());
+
+		if amount.is_zero() {
+			return Err(ContractError::ZeroTransferAmount)
+		}
+
+		let transfer_amount = {
+			if amount.slope.0 == 0 {
+				amount.intercept.0
+			} else {
+				let query_msg = Cw20QueryMsg::Balance { address: to.clone() };
+				let response: BalanceResponse =
+					deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+						contract_addr: cw20_address.addr.clone().into_string(),
+						msg: to_binary(&query_msg)?,
+					}))?;
+				amount.apply(response.balance.into())
```

Comment:
Why are we checking the slope instead of applying `apply` in both cases?

Reply:
Yeah actually this is unnecessary.